### ISSUE TITLE
Recommended way of using more than one sharded dataset

### ISSUE THREAD

**Issue by mchalecki** [tmbdev/webdataset#19]
I have few big datasets which I want to shard and use in one dataloader.

1. Can I use safely ConcatDataset ?
2. Should I use ConcatDataset or is there an url pattern that I can use ?

**Comment by tmbdev** [comment]
The `webdataset.Dataset` class takes shards either as a string or a list of URLs as an argument. If you give it a string, it simply expands it for you using braceexpand as a convenience.

You can construct the list of shards any way you like and just pass it as a list:

```Python
urls = list(braceexpand(dataset1)) + list(braceexpand(dataset2))
dataset.= webdataset.Dataset(urls) 
```

**Comment by tmbdev** [comment]
TODO: add documentation strings

**Comment by srinidhiPY** [comment]
Hello, First I would like to thank all developers for their efforts in developing this wonderful library for our community. Keep it up!

My issue: How do we load multiple webdatasets into single pytorch data loader, while ensuring the sampling order is preserved between them.

Let's take an example: I have two web datasets with it's contents in the form

1) Dataset1.tar
- cat11.png
-cat11.cls
-dog11.png
-dog11.cls

2) Dataset2.tar
-cat21.png
-cat21.cls
-dog21.png
-dog21.cls

I need to load webdatasets (Dataset1 and 2) into single pytorch data loader. Also, during the iteration over the batch I need to preserve their order for training the network. 

Example: output

For batch_id, (input1, class1, input2, class2) in enumerate(loader):
Input1 = cat11.png
Class1 = Cat11.cls
Input2 = cat21.png
Class2 = Cat21.cls

Please,  kindly suggest me how to obtain this format? Thanks and greatly appreciate for this help.

**Comment by tmbdev** [comment]
I'm not sure I fully understand your example. Why would you want "cat11.png" and "cat21.png" to have their order preserved? Aren't they unrelated training samples?

Relying on the order of samples in a dataset tends to cause data management issues in any DL framework; if you depend on two images being present during training in a particular order, it's best to put them into the same sample. In WebDataset, you do that just by file name. Here are two training samples (cat1 and cat2) each consisting of two images (1.png and 2.png):

```Python
cat1.1.png
cat1.1.cls
cat1.2.png
cat1.2.cls
cat2.1.png
cat2.1.cls
cat2.2.png
cat2.2.cls
...
```

The second part of your question is about merging two datasets. If you don't depend on order, you can simply list all the shards you want explicitly in a Python list:

```Python
import braceexpand
allshards = braceexpand.braceexpand("dataset1-{0000..9999}.tar") + braceexpand.braceexpand("dataset2-{0000..9999}.tar")
dataset = WebDataset(allshards)
```

If you really want an order-preserving merge of shards during training, you can write a small helper class to do so; but, again, while WebDataset is happy to do this for you, it's not a good idea from a data management perspective:

```Python
from webdataset import Composable, Shorthands  # not needed, just for syntactic convenience

class RoundRobin(IterableDataset, Composable, Shorthands):
    def __init__(self, datasets):
        self.datasets = datasets
    def __iter__(self):
        sources = [iter(ds) for ds in self.dataset]
        while True:
            for source in sources:
                yield next(source)
```

**Comment by srinidhiPY** [comment]
Hi there,
Thank you so much for your kind support.

Sorry my earlier explanation was much confusing. But, after your
suggestion, I have rearranged my dataset in a folder, which is as follows:

Dataset.tar - folder which contains image and class labels
Cat.aug1.png
Cat.aug2.png
Cat.aug3.png
Cat.cls
Dog.aug1.png
Dog.aug2.png
Dog.aug3.png
Dog.cls

The samples: Cat.aug1, Cat.aug2, Cat.aug3. are the different augmented view
of the same input image with label Cat.cls

I am interested in feeding all three images (Cat.aug1, Cat.aug2, Cat.aug3)
at the same time to a neural network with label. I am wondering how to
shuffle this dataset and still extract the corresponding 3 samples
belonging to the same input image (cat/dog) using web dataset. Please, let
me know how to deal with this.

Thank you so much for your time and patience.
Chetan
